scene/resources/mesh.cpp

```diff
@@ -236,7 +236,22 @@ Ref<Shape> Mesh::create_convex_shape() const {
 		Array a = surface_get_arrays(i);
 		ERR_FAIL_COND_V(a.empty(), Ref<ConvexPolygonShape>());
 		PoolVector<Vector3> v = a[ARRAY_VERTEX];
-		vertices.append_array(v);
+		int vert_count = v.size();
+		PoolVector<Vector3>::Read vr = v.read();
+		// Insert verts only if they're not duplicates.
+		for (int insert_i = 0; insert_i < vert_count; ++insert_i) {
+			Vector3 vert = vr[insert_i];
+			bool duplicate = false;
+			for (int vi = 0; vi < vertices.size(); ++vi) {
+				if (vertices[vi].is_equal_approx(vert)) {
+					duplicate = true;
+					break;
+				}
+			}
+			if (!duplicate) {
+				vertices.append(vert);
+			}
+		}
```

Comment:
This loop has O(n²) runtime.
Input meshes may easily have one million or more vertices, which would lead to a very long hang in the editor.

Reply:
> 
> 
> This loop has O(n²) runtime.
> Input meshes may easily have one million or more vertices, which would lead to a very long hang in the editor.

I would hope somebody isn't trying to use a million+vert mesh as convex collision.  This should just be used for simple shapes.  I think a hang in the editor would be the least of your problems if you tried that.  Do you have a suggestion for an alternative algorithm to check if a vertex already exists before adding it?

Reply:
We could keep track of the time spent in generation and abort if > 10 seconds have passed, then display an error dialog.

Reply:
I wouldn't even remove the double vertices. That's up to the convex hull algorithm, which should be O(n log n).
I'd say we compute the convex hull right away, or we just leave the data as it is passed in.

But if you really want to remove duplicates, I'd just use a (Hash)set. That works fine if the coordinates are exactly the same.
Efficiently removing doubles with a tolerance is more complicated. I think Blender builds a KD-Tree for that.

Reply:
You could use a set with a custom comparator to check for duplicated vertices with a tolerance.
https://github.com/godotengine/godot/blob/master/core/templates/set.h

The default comparator is this one:
https://github.com/godotengine/godot/blob/6f8d5cff66416993856bdb01c83b999b84425fb0/core/typedefs.h#L232-L236

This comparator should do the trick (I haven't tested though):
```
struct Comparator { 
    bool operator()(const Vector3 &p_a, const Vector3 &p_b) const
    {
        if (Math::is_equal_approx(p_a.x, p_b.x)) {
		if (Math::is_equal_approx(p_a.y == p_b.y)) {
			return p_a.z < p_b.z - CMP_EPSILON;
		}
		return p_a.y < p_b.y - CMP_EPSILON;
	}
	return p_a.x < p_b.x - CMP_EPSILON;
    } 
};
```

Reply:
The tolerance will only work in one dimension like that. For more dimensions, you need something like a KD tree.

Reply:
@mortarroad I'm not sure I understand what you mean by working only in one dimension but I'm interested to learn. Is there more explanation or an external source you can share?

In any case, I wonder now if the best solution could be to just use VHACD for single collision shapes, since we're already using it in the editor for multiple convex shapes. We can just set it to one convex max for this case. It does a very good job at simplifying vertices without messing with the volume.

It would need a bit more testing, but when I did a quick test while looking into #48468, it seemed to work fine this way.

Reply:
I don't have any references, but I can give you an example.
For this example, let's say the tolerance `DUPLICATE_VERT_EPSILON` is such that two numbers that are an integer are apart are unequal. (e.g. 1 and 2 are different; 1 and 1.9 are equal). E.g. `#define DUPLICATE_VERT_EPSILON 0.99`

Now insert the vectors (0, 0, 0), (-0.9, 1, 0), and (-1.1, 1, 0); in this order. This yields the following tree:
```
             (0, 0, 0)
             /       \
 (-1.1, 1, 0)        (-0.9, 1, 0)
 /          \        /          \
```
(-0.9, 1, 0) is inserted right of (0, 0, 0), because the x compares equal, therefore the y coordinate determines that it is greater.
(-1.1, 1, 0) is inserted left of (0, 0, 0), because -1.1 does not compare equal to 0, and then compares as less, even with the tolerance.

However, (-1.1, 1, 0) and (-0.9, 1, 0) would compare equal; they are only 0.2 apart! But they never get compared with each other.


Yes, the VHACD approach looks interesting.

Reply:
Sorry, I thought I replied last time. Thanks for your detailed explanation, it makes a lot of sense!